File: easy/3622-check-divisibility-by-digit-sum-and-product/solution.py
class Solution:
    def checkDivisibility(self, n: int) -> bool:
        """
        3622. Check Divisibility by Digit Sum and Product
        Time: O(d)
        Space: O(1)
        """
        digit_sum = 0
        digit_product = 1
        x = n
        while x > 0:
            d = x % 10
            digit_sum += d
            digit_product *= d
            x //= 10

        # digit_sum >= 1 for any positive n, so the divisor is never zero
        return n % (digit_sum + digit_product) == 0

File: easy/3622-check-divisibility-by-digit-sum-and-product/solution.java
/**
 * 3622. Check Divisibility by Digit Sum and Product
 * Time: O(d)
 * Space: O(1)
 */
class Solution {
    public boolean checkDivisibility(int n) {
        int digitSum = 0;
        int digitProduct = 1;
        for (int x = n; x > 0; x /= 10) {
            int d = x % 10;
            digitSum += d;
            digitProduct *= d;
        }

        // digitSum >= 1 for any positive n, so the divisor is never zero
        return n % (digitSum + digitProduct) == 0;
    }
}
